File: 007 有序表的合并/使用数组实现.c
#include <stdio.h>

void merge(int *arrA, int lenA, int *arrB, int lenB, int *arrC) {
    int idxA = 0;
    int idxB = 0;
    
    while (idxA < lenA && idxB < lenB) {
        int valA = arrA[idxA];
        int valB = arrB[idxB];
        
        if (valA < valB) {
            arrC[idxA + idxB] = valA;
            idxA++;
        } else {
            arrC[idxA + idxB] = valB;
            idxB++;
        }
    }
    
    while (idxA < lenA) {
        arrC[idxA + idxB] = arrA[idxA];
        idxA++;
    }
    while (idxB < lenB) {
        arrC[idxA + idxB] = arrB[idxB];
        idxB++;
    }
}

void traverse(int *array, int length) {
    for (int i = 0; i < length; i++) {
        printf("%d ", array[i]);
    }
    printf("\n");
}

int main() {
    int arrA[4] = {21, 32, 36, 88};
    int arrB[7] = {2, 15, 37, 99, 123, 345, 432};
    int arrC[11] = {};
    
    
    printf("数组A：");
    traverse(arrA, 4);
    printf("数组B：");
    traverse(arrB, 7);
    
    
    merge(arrA, 4, arrB, 7, arrC);
    printf("合并后数组：");
    traverse(arrC, 11);
    
    return 0;
}

File: 007 有序表的合并/使用数组实现.swift
import Foundation

func merge(_ listA: [Int], _ listB: [Int]) -> [Int] {
    let lenA = listA.count
    let lenB = listB.count
    
    var listC = Array<Int>(repeating: 0, count: lenA + lenB)
    
    var idxA = 0
    var idxB = 0
    
    while idxA < lenA && idxB < lenB {
        let valA = listA[idxA]
        let valB = listB[idxB]
        
        if valA < valB {
            listC[idxA + idxB] = valA
            idxA += 1
        } else {
            listC[idxA + idxB] = valB
            idxB += 1
        }
    }
    
    while idxA < lenA {
        listC[idxA + idxB] = listA[idxA]
        idxA += 1
    }
    
    while idxB < lenB {
        listC[idxA + idxB] = listB[idxB]
        idxB += 1
    }
    
    return listC
}

func traverse(_ array: [Int]) {
    for v in array {
        print(v, terminator: " ")
    }
    print()
}

// 测试数据
var arrA = [Int]()
var arrB = [Int]()

for _ in 0..<6 {
    arrA.append(Int.random(in: 0..<30))
}

for _ in 0..<9 {
    arrB.append(Int.random(in: 0..<90))
}

let listA = arrA.sorted()
let listB = arrB.sorted()

print("数组A：")
traverse(listA)
print("数组B：")
traverse(listB)

print("合并后数组：")
let listC = merge(listA, listB)
traverse(listC)
